nest/src/BoringVaultPredeposit.sol

```diff
@@ -0,0 +1,851 @@
+
+    /**
+     * @notice Indicates a failure because the user does not have enough tokens staked
+     * @param user Address of the user who does not have enough tokens staked
+     * @param token Token contract address
+     * @param amount Amount of tokens that the user wants to withdraw
+     * @param amountStaked Amount of tokens that the user has staked
+     */
+    error InsufficientStaked(address user, IERC20 token, uint256 amount, uint256 amountStaked);
+
+    /// @notice Thrown when a user tries to convert to vault shares before the conversion start time
+    /// @param currentTime The current block timestamp
+    /// @param startTime The configured vault conversion start time
+    error ConversionNotStarted(uint256 currentTime, uint256 startTime);
+
+    // Modifiers
+
+    /// @notice Only the timelock contract can call this function
+    modifier onlyTimelock() {
+        if (msg.sender != address(_getBoringVaultPredepositStorage().timelock)) {
+            revert Unauthorized(msg.sender, address(_getBoringVaultPredepositStorage().timelock));
+        }
+        _;
+    }
+
+    // Initializer
+
+    /**
+     * @notice Prevent the implementation contract from being initialized or reinitialized
+     * @custom:oz-upgrades-unsafe-allow constructor
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initialize the contract.
+     * @param timelock The timelock contract address.
+     * @param owner The owner address.
+     * @param boringVaultConfig The configuration for the BoringVault.
+     * @param salt A value (e.g. derived from the vault’s parameters) used to derive a unique storage slot.
+     *
+     * The storage slot is computed as:
+     * keccak256(abi.encodePacked("plume.storage.BoringVaultPredeposit", salt))
+     */
+    function initialize(
+        TimelockController timelock,
+        address owner,
+        BoringVault memory boringVaultConfig,
+        bytes32 salt
+    ) public initializer {
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+        __ReentrancyGuard_init();
+
+        _storageSlot = keccak256(abi.encodePacked("plume.storage.BoringVaultPredeposit", salt));
+        BoringVaultPredepositStorage storage $ = _getBoringVaultPredepositStorage();
+
+        $.multisig = owner;
+        $.timelock = timelock;
+        $.vault = boringVaultConfig;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, owner);
+        _grantRole(ADMIN_ROLE, owner);
+    }
+
+    /**
+     * @notice Reinitialize the BoringVaultPredeposit contract by adding the timelock and multisig contract address
+     * @param multisig Multisig contract address
+     * @param timelock Timelock contract address
+     */
+    function reinitialize(address multisig, TimelockController timelock) public reinitializer(2) onlyRole(ADMIN_ROLE) {
+        BoringVaultPredepositStorage storage $ = _getBoringVaultPredepositStorage();
+        $.multisig = multisig;
+        $.timelock = timelock;
```

Comment:
At this point, the contract's super admin role and admin role is still the "owner", potentially the deployer.

All the following functions requires "ADMIN_ROLE" which has nothing to do with this multisig we set here.

If I understand our intention correctly, we must revoke and grant roles with reinitialize(), and emitting events accordingly:

```

    // Transfer administrative roles
    address oldAdmin = getRoleMember(DEFAULT_ADMIN_ROLE, 0);
    _revokeRole(DEFAULT_ADMIN_ROLE, oldAdmin);
    _revokeRole(ADMIN_ROLE, oldAdmin);
    
    _grantRole(DEFAULT_ADMIN_ROLE, multisig);
    _grantRole(ADMIN_ROLE, multisig);
    
    emit MultisigSet(multisig);
    emit TimelockSet(address(timelock));
```